File: C++/Arrays/Counting_sort.cpp
#include <iostream>
using namespace std;

int main() {
    int input_arr[] = {1,2,1,1,3,4,0,5,6,1,0,1};
    int n = sizeof(input_arr)/sizeof(input_arr[0]);
    int max = 0;
    for (int i = 0; i < n; i++)
        max = input_arr[i] >= max ? input_arr[i] : max;
    
    int count_arr[max + 1]; // count array is of size max + 1
    for (int i = 0; i <= max; i++) // initialize count array with zeros
        count_arr[i] = 0;

    for (int i = 0; i < n; i++) // count occurrences of each number
        count_arr[input_arr[i]] += 1;
    
    for (int i = 1; i <= max; i++) // make cumulative array
        count_arr[i] += count_arr[i - 1];

    int output_arr[n];
    for (int i = 0; i < n; i++){ // output array
        output_arr[count_arr[input_arr[i]] - 1] = input_arr[i];
        count_arr[input_arr[i]] -= 1;
    }
    for (int i = 0; i < n; i++)
        cout << output_arr[i] << " ";
}

File: Python/Arrays/Counting_sort.py
arr = [1, 2, 1, 1, 3, 4, 0, 5, 6, 1, 0, 1]
n = len(arr)

# find max val
max_val = 0
for i in arr:
    max_val = max(max_val, i)

# count array
count_arr = [0] * (max_val + 1) # initialize with 0
for i in arr:
    count_arr[i] += 1

# make cumulative array
for i in range(1, max_val + 1):
    count_arr[i] += count_arr[i - 1]

output_arr = [0] * n
for i in arr:
    output_arr[count_arr[i] - 1] = i
    count_arr[i] -= 1

for i in output_arr:
    print(i, end = ' ')